lib/ldp.js

```diff
@@ -478,27 +481,18 @@ class LDP {
     })
   }
 
-  getAvailablePath (host, containerURI, slug, callback) {
-    var self = this
-    slug = slug || uuid.v1()
-
-    function ensureNotExists (newPath) {
-      return new Promise(resolve => {
-        self.exists(host, newPath, function (err) {
-          // If an error occurred, the resource does not exist yet
-          if (err) {
-            resolve(newPath)
-          // Otherwise, generate a new path
-          } else {
-            const id = uuid.v1().split('-')[ 0 ] + '-'
-            newPath = path.join(containerURI, id + slug)
-            resolve(ensureNotExists(newPath))
-          }
-        })
-      })
+  getAvailablePath (host, containerURI, { slug = uuid.v1(), extension }) {
+    function ensureNotExists (self, newPath) {
```

Comment:
I'd refactor this closure into a function fully determined by its arguments, that way it doesn't have to be redefined at runtime.

Reply:
It's a bit more tricky, since the `ensureNotExists` function is calling itself recursively. Basically, the flow is that it receives a filename, and if that name exists, generates another name (which it needs to check again).

Note that "redefined" is very likely not how V8 interprets this. The function exists once as machine code and will just be populated with different arguments.

Reply:
I was trying to think this through. I wish I knew better how JIT compilers work (/me makes note to do more reading on this subject). My hypothesis was that, because `ensureNotExists` depends on the external values of the parameters to `getAvailablePath` (`host`, `containerURI`, `slug`, and `extension`), it cannot be translated into machine code, because its very definition is determined at runtime.

That, or I have no idea what I'm talking about 😆